bayes_kit/metropolis.py

```diff
@@ -0,0 +1,153 @@
+    adjusted_log_acceptance_ratio = (
+        log_acceptance_ratio + log_transition_likelihood_ratio
+    )
+    log_uniform_random: float = np.log(rand.uniform())
+    return log_uniform_random < adjusted_log_acceptance_ratio
+
+
+class MetropolisHastings:
+    def __init__(
+        self,
+        model: LogDensityModel,
+        proposal_fn: Callable[[NDArray[np.float64]], ArrayLike],
+        transition_lp_fn: TransitionLPFn,
+        *,
+        init: Optional[NDArray[np.float64]] = None,
+        seed: Union[None, int, np.random.BitGenerator, np.random.Generator] = None,
+    ):
+        self._model = model
+        self._dim = self._model.dims()
+        self._rand = np.random.default_rng(seed)
+        self._proposal_fn = (
+            proposal_fn  # TODO: type check this once during construction?
+        )
+        self._transition_lp_fn = transition_lp_fn
+        self._theta = init or self._rand.normal(size=self._dim)
+        self._log_p_theta = self._model.log_density(self._theta)
+
+    def __iter__(self) -> Iterator[Sample]:
+        return self
+
+    def __next__(self) -> Sample:
+        return self.sample()
+
+    def sample(self) -> Sample:
+        proposal, lp_proposal = self._propose()
+        accepted = self._accept_test(lp_proposal, proposal)
+        if accepted:
+            self._update_theta(proposal, lp_proposal)
+        return (self._theta, self._log_p_theta)
+
+    def _propose(self) -> Sample:
+        untyped_proposed_theta = np.asanyarray(self._proposal_fn(self._theta))
+        if not (untyped_proposed_theta.dtype == np.float64):
+            raise TypeError(
+                f"Proposal function must return a double; returned {untyped_proposed_theta.dtype=}."
+            )
+        proposed_theta: NDArray[np.float64] = untyped_proposed_theta
+        lp_proposed_theta = self._model.log_density(proposed_theta)
+        return (proposed_theta, lp_proposed_theta)
+
+    def _update_theta(self, theta: NDArray[np.float64], log_p_theta: float) -> None:
+        self._theta = theta
+        self._log_p_theta = log_p_theta
+
+    def _accept_test(self, lp_proposal: float, proposal: NDArray[np.float64]) -> bool:
+        lp_forward_transition = self._transition_lp_fn(self._theta, proposal)
+        lp_reverse_transition = self._transition_lp_fn(proposal, self._theta)
+        return metropolis_hastings_accept_test(
+            lp_proposal,
+            self._log_p_theta,
+            lp_forward_transition,
+            lp_reverse_transition,
+            self._rand,
+        )
+
+
+class Metropolis(MetropolisHastings):
+    def __init__(
+        self,
+        model: LogDensityModel,
+        proposal_fn: Callable[[NDArray[np.float64]], ArrayLike],
+        *,
+        init: Optional[NDArray[np.float64]] = None,
+        seed: Union[None, int, np.random.BitGenerator, np.random.Generator] = None,
+    ):
+        dummy_transition_fn: TransitionLPFn = lambda a, b: 1
```

Comment:
By overriding `_accept_test`, this lambda isn't actually ever called, right? Just checking (and it is probably worth documening if so)